crux-calcite/src/crux/calcite.clj

```diff
@@ -0,0 +1,180 @@
+  [schema ^RexNode filter*]
+  (condp = (.getKind filter*)
+    SqlKind/EQUALS
+    [[(apply list '= (->operands schema filter*))]]
+    SqlKind/NOT_EQUALS
+    [(list 'not [(apply list '= (->operands schema filter*))])]
+    SqlKind/AND
+    (mapcat (partial ->crux-where-clauses schema) (.-operands ^RexCall filter*))
+    SqlKind/OR
+    [(apply list 'or (mapcat (partial ->crux-where-clauses schema) (.-operands ^RexCall filter*)))]
+    SqlKind/INPUT_REF
+    [[(list '= (operand->v filter* schema) true)]]
+    SqlKind/NOT
+    [(apply list 'not (mapcat (partial ->crux-where-clauses schema) (.-operands ^RexCall filter*)))]
+    SqlKind/GREATER_THAN
+    [[(list '> (->operands schema filter*))]]
+    SqlKind/GREATER_THAN_OR_EQUAL
+    [[(apply list '>= (->operands schema filter*))]]
+    SqlKind/LESS_THAN
+    [[(apply list '< (->operands schema filter*))]]
+    SqlKind/LESS_THAN_OR_EQUAL
+    [[(apply list '<= (->operands schema filter*))]]
+    SqlKind/LIKE
+    [[(apply list 'like (->operands schema filter*))]]
+    SqlKind/IS_NULL
+    [[(apply list '= (->operands schema filter*))]]
+    SqlKind/IS_NOT_NULL
+    [[(list 'not (apply list '= (->operands schema filter*)))]]))
+
+(defn- ->crux-query
+  [schema filters projects]
+  (try
+    (let [{:keys [crux.sql.table/columns crux.sql.table/query]} schema
+          {:keys [find where]} query]
+      {:find (if (seq projects) (mapv find projects) find)
+       :where (vec
+               (concat
+                (mapcat (partial ->crux-where-clauses schema) filters)
+                where))
+       :args [{:like #(org.apache.calcite.runtime.SqlFunctions/like %1 %2)}]})
+    (catch Throwable e
+      (log/error e)
+      (throw e))))
+
+(def ^:private column-types->sql-types {:varchar SqlTypeName/VARCHAR
+                                        :keyword SqlTypeName/VARCHAR
+                                        :integer SqlTypeName/INTEGER
+                                        :long SqlTypeName/BIGINT
+                                        :boolean SqlTypeName/BOOLEAN
+                                        :double SqlTypeName/DOUBLE
+                                        :datetime SqlTypeName/DATE})
+
+(defn- ^java.util.List perform-query [node q]
+  (->> (crux/q (crux/db node) q)
+       (mapv to-array)))
+
+(defn- ^String ->column-name [c]
+  (string/replace (string/upper-case (str c)) #"^\?" ""))
+
+(defn- make-table [node {:keys [:crux.sql.table/query :crux.sql.table/columns] :as table-schema}]
+  (proxy
+      [org.apache.calcite.schema.impl.AbstractTable
+       org.apache.calcite.schema.ProjectableFilterableTable]
+      []
+      (getRowType [^RelDataTypeFactory type-factory]
+        (let [field-info  (RelDataTypeFactory$Builder. type-factory)]
+          (doseq [c (:find query)]
+            (let  [col-name (->column-name c)
+                   col-type ^SqlTypeName (column-types->sql-types (columns c))]
+              (assert col-type (str "Unrecognized column: " c))
```

Comment:
this is an error I'd want to throw in production (because it's checking user input), and we can't rely on assertions being turned on - could this be an explicit `(or ... (throw (IllegalArgumentException. ...))`?